godef_test.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"go/build"
+	"go/token"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/rogpeppe/godef/go/types"
+	"golang.org/x/tools/go/packages/packagestest"
+)
+
+func TestGoDef(t *testing.T) { packagestest.TestAll(t, testGoDef) }
+func testGoDef(t *testing.T, exporter packagestest.Exporter) {
+	const godefAction = ">"
+	modules := []packagestest.Module{{
+		Name: "github.com/rogpeppe/godef",
+		Files: map[string]interface{}{
+			"a/a.go":      packagestest.Copy("testdata/a/a.go"),
+			"a/random.go": packagestest.Copy("testdata/a/random.go"),
+			"b/b.go":      packagestest.Copy("testdata/b/b.go"),
+		},
+	}}
+	exported := packagestest.Export(t, exporter, modules)
+	defer exported.Cleanup()
+
+	posStr := func(p token.Position) string {
+		return localPos(p, exported, modules)
+	}
+
+	const gopathPrefix = "GOPATH="
+	const gorootPrefix = "GOROOT="
+	for _, v := range exported.Config.Env {
+		if strings.HasPrefix(v, gopathPrefix) {
+			build.Default.GOPATH = v[len(gopathPrefix):]
+		}
+		if strings.HasPrefix(v, gorootPrefix) {
+			build.Default.GOROOT = v[len(gorootPrefix):]
+		}
+	}
+
+	count := 0
+	exported.Expect(map[string]interface{}{
+		"godef": func(src, target token.Position) {
+			count++
+			input, err := ioutil.ReadFile(src.Filename)
+			if err != nil {
+				t.Errorf("Failed %v: %v", src, err)
+				return
+			}
+			obj, _, err := godef(src.Filename, input, src.Offset)
+			if err != nil {
+				t.Errorf("Failed %v: %v", src, err)
+				return
+			}
+			pos := types.FileSet.Position(types.DeclPos(obj))
+			check := token.Position{
+				Filename: pos.Filename,
+				Line:     pos.Line,
+				Column:   pos.Column,
+				Offset:   pos.Offset,
+			}
+			if posStr(check) != posStr(target) {
+				t.Errorf("Got %v expected %v", posStr(check), posStr(target))
+			}
+		},
+	})
+	if count == 0 {
+		t.Fatalf("No godef tests were run")
+	}
+}
+
+var cwd, _ = os.Getwd()
+
+func localPos(pos token.Position, e *packagestest.Exported, modules []packagestest.Module) string {
+	fstat, fstatErr := os.Stat(pos.Filename)
+	if fstatErr == nil {
+		for _, m := range modules {
+			for fragment := range m.Files {
+				fname := e.File(m.Name, fragment)
+				if s, err := os.Stat(fname); err == nil && os.SameFile(s, fstat) {
+					pos.Filename = filepath.Join(cwd, "testdata", filepath.FromSlash(fragment))
+					break
```

Comment:
break out of all the loops? (or return here?)

Reply:
Done